App fails to start if installed from a signed APK/AAB

The app works absolutely fine if I install it from an unsigned APK via Android Studio. However, if installed from a signed AAB/APK (as would be the case if downloaded from the Google Play Store), it'd crash as soon as it starts.

I've already tried a brand new keyring, and cleared the cache and `build` folders.

The issue appears to be related to the desugaring functionality in the Android Gradle plugin. I suspect Android Studio isn't executing the appropriate tasks when generating signed APKs/AABs (most likely due to a configuration issue). Here's the traceback:

```
2021-04-09 17:01:06.141 10359-10374/? E/AndroidRuntime: FATAL EXCEPTION: DefaultDispatcher-worker-1
    Process: tech.relaycorp.ping, PID: 10359
    java.lang.NoClassDefFoundError: Failed resolution of: Lj$/$r8$retargetLibraryMember$virtualDispatch$Date$toInstant$dispatchHolder;
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.dateToZonedDateTime(Certificate.kt:321)
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.getStartDate(Certificate.kt:170)
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.validateValidityPeriod(Certificate.kt:224)
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.validate(Certificate.kt:218)
        at tech.relaycorp.awaladroid.endpoint.PublicThirdPartyEndpoint$Companion.import(ThirdPartyEndpoint.kt:179)
        at tech.relaycorp.ping.domain.AddPublicPeer.add(AddPublicPeer.kt:17)
        at tech.relaycorp.ping.domain.BootstrapData.importDefaultPublicPeer(BootstrapData.kt:27)
        at tech.relaycorp.ping.domain.BootstrapData.bootstrapIfNeeded(BootstrapData.kt:21)
        at tech.relaycorp.ping.domain.BootstrapData$bootstrapIfNeeded$1.invokeSuspend(Unknown Source:11)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "j$.$r8$retargetLibraryMember$virtualDispatch$Date$toInstant$dispatchHolder" on path: DexPathList[[zip file "/data/app/tech.relaycorp.ping-Qv974gcM15UExBdf4eRz9g==/base.apk", zip file "/data/app/tech.relaycorp.ping-Qv974gcM15UExBdf4eRz9g==/split_config.en.apk", zip file "/data/app/tech.relaycorp.ping-Qv974gcM15UExBdf4eRz9g==/split_config.xxhdpi.apk"],nativeLibraryDirectories=[/data/app/tech.relaycorp.ping-Qv974gcM15UExBdf4eRz9g==/lib/arm64, /system/lib64, /vendor/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:125)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:379)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:312)
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.dateToZonedDateTime(Certificate.kt:321) 
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.getStartDate(Certificate.kt:170) 
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.validateValidityPeriod(Certificate.kt:224) 
        at tech.relaycorp.relaynet.wrappers.x509.Certificate.validate(Certificate.kt:218) 
        at tech.relaycorp.awaladroid.endpoint.PublicThirdPartyEndpoint$Companion.import(ThirdPartyEndpoint.kt:179) 
        at tech.relaycorp.ping.domain.AddPublicPeer.add(AddPublicPeer.kt:17) 
        at tech.relaycorp.ping.domain.BootstrapData.importDefaultPublicPeer(BootstrapData.kt:27) 
        at tech.relaycorp.ping.domain.BootstrapData.bootstrapIfNeeded(BootstrapData.kt:21) 
        at tech.relaycorp.ping.domain.BootstrapData$bootstrapIfNeeded$1.invokeSuspend(Unknown Source:11) 
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678) 
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665) 
    	Suppressed: java.io.IOException: No original dex files found for dex location /data/app/tech.relaycorp.ping-Qv974gcM15UExBdf4eRz9g==/split_config.en.apk
        at dalvik.system.DexFile.openDexFileNative(Native Method)
        at dalvik.system.DexFile.openDexFile(DexFile.java:353)
        at dalvik.system.DexFile.<init>(DexFile.java:100)
        at dalvik.system.DexFile.<init>(DexFile.java:74)
        at dalvik.system.DexPathList.loadDexFile(DexPathList.java:374)
        at dalvik.system.DexPathList.makeDexElements(DexPathList.java:337)
        at dalvik.system.DexPathList.<init>(DexPathList.java:157)
        at dalvik.system.BaseDexClassLoader.<init>(BaseDexClassLoader.java:65)
        at dalvik.system.PathClassLoader.<init>(PathClassLoader.java:64)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:73)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:88)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:69)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:35)
        at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:693)
        at android.app.LoadedApk.getClassLoader(LoadedApk.java:727)
        at android.app.LoadedApk.getResources(LoadedApk.java:954)
        at android.app.ContextImpl.createAppContext(ContextImpl.java:2270)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5639)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1656)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
2021-04-09 17:01:06.142 10359-10374/? E/AndroidRuntime:     at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
    	Suppressed: java.io.IOException: No original dex files found for dex location /data/app/tech.relaycorp.ping-Qv974gcM15UExBdf4eRz9g==/split_config.xxhdpi.apk
        at dalvik.system.DexFile.openDexFileNative(Native Method)
        at dalvik.system.DexFile.openDexFile(DexFile.java:353)
        at dalvik.system.DexFile.<init>(DexFile.java:100)
        at dalvik.system.DexFile.<init>(DexFile.java:74)
        at dalvik.system.DexPathList.loadDexFile(DexPathList.java:374)
        at dalvik.system.DexPathList.makeDexElements(DexPathList.java:337)
        at dalvik.system.DexPathList.<init>(DexPathList.java:157)
        at dalvik.system.BaseDexClassLoader.<init>(BaseDexClassLoader.java:65)
        at dalvik.system.PathClassLoader.<init>(PathClassLoader.java:64)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:73)
        at com.android.internal.os.ClassLoaderFactory.createClassLoader(ClassLoaderFactory.java:88)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:69)
        at android.app.ApplicationLoaders.getClassLoader(ApplicationLoaders.java:35)
        at android.app.LoadedApk.createOrUpdateClassLoaderLocked(LoadedApk.java:693)
        at android.app.LoadedApk.getClassLoader(LoadedApk.java:727)
        at android.app.LoadedApk.getResources(LoadedApk.java:954)
        at android.app.ContextImpl.createAppContext(ContextImpl.java:2270)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5639)
        at android.app.ActivityThread.-wrap1(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1656)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

@sdsantos, can you please try reproducing this? There's no need to upload anything to the Google Play Store: Just generate a local key and generate a signed APK: https://developer.android.com/studio/publish/app-signing#sign-apk

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -37,7 +37,9 @@ android {
         }
     }
     compileOptions {
-        coreLibraryDesugaringEnabled true
+        // Workaround for https://github.com/relaycorp/awala-ping-android/issues/39
+        // coreLibraryDesugaringEnabled true
+
         sourceCompatibility = JavaVersion.VERSION_1_8
         targetCompatibility = JavaVersion.VERSION_1_8
     }

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -15,7 +15,7 @@ android {
 
     defaultConfig {
         applicationId "tech.relaycorp.ping"
-        minSdkVersion 21
+        minSdkVersion 24
         targetSdkVersion 30
         versionCode 2
         versionName "0.2"

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -15,7 +15,7 @@ android {
 
     defaultConfig {
         applicationId "tech.relaycorp.ping"
-        minSdkVersion 24
+        minSdkVersion 26
         targetSdkVersion 30
         versionCode 2
         versionName "0.2"

--- a/app/firebase-test-lab.yml
+++ b/app/firebase-test-lab.yml
@@ -4,17 +4,18 @@ spec:
   test: app/build/outputs/apk/androidTest/debug/app-debug-androidTest.apk
   timeout: 30m
   device:
-    # Asus Nexus 7 (physical). The oldest device we're testing.
-    - model: flo
-      version: 21   # Android 5.0
-      locale: en
-      orientation: portrait
-
-    # Nexus 5X (virtual)
-    - model: Nexus5X
-      version: 24   # Android 7.0
-      locale: en
-      orientation: portrait
+# TODO: Test on Android API < 26 once we restore support for it.
+#    # Asus Nexus 7 (physical). The oldest device we're testing.
+#    - model: flo
+#      version: 21   # Android 5.0
+#      locale: en
+#      orientation: portrait
+#
+#    # Nexus 5X (virtual)
+#    - model: Nexus5X
+#      version: 24   # Android 7.0
+#      locale: en
+#      orientation: portrait
 
     # Nexus 5X (virtual)
     - model: Nexus5X

--- a/app/lint.xml
+++ b/app/lint.xml
@@ -10,4 +10,7 @@
     <issue id="SelectableText" severity="ignore" />
     <issue id="UnusedIds" severity="informational" />
     <issue id="UnusedResources" severity="informational" />
+
+    <!-- TODO: Remove once we restore support for Android API < 26 -->
+    <issue id="ObsoleteSdkInt" severity="informational" />
 </lint>

